### PR TITLE
SQL Parsing

### PR DESCRIPTION
**Pull request by shehabgamin**

### CODE REVIEW

File: crates/sail-spark-connect/tests/gold_data/expression/interval.json [hunk 1]
```diff
@@ -3589,13 +3623,29 @@
       "input": "INTERVAL '1 year 2 second'",
       "exception": "\nCannot mix year-month and day-time fields: INTERVAL '1 year 2 second'.(line 1, pos 0)\n\n== SQL ==\nINTERVAL '1 year 2 second'\n^^^\n",
```
Comment (shehabgamin) [inline crates/sail-spark-connect/tests/gold_data/expression/interval.json:3624]: We should remove this exception @linhr

File: crates/sail-common/src/spec/data_type.rs [hunk 2]
```diff
@@ -31,7 +33,9 @@ pub enum DataType {
         length: u32,
     },
     Date,
-    Timestamp,
+    // TODO: This is the preliminary work towards Sail spec::DataType achieving parity with Arrow.
+    //  spec::DataType::Timestamp and TimestampNtz need to be merged in follow-up work.
```
Comment (linhr) [inline crates/sail-common/src/spec/data_type.rs]: As discussed offline, we still need `TimestampNtz` (and possibly `TimestampLtz`). So let's remove this line of the comment.

File: crates/sail-common/src/spec/literal.rs [hunk 3]
```diff
@@ -21,8 +23,13 @@ pub enum Literal {
     Date {
         days: i32,
     },
-    Timestamp {
+    // TODO: This is the preliminary work towards Sail spec::Literal achieving parity with Arrow.
+    //  spec::Literal::Timestamp and TimestampNtz need to be merged in follow-up work.
```
Comment (linhr) [inline crates/sail-common/src/spec/literal.rs]: Remove this line.

File: crates/sail-plan/src/function/aggregate.rs [hunk 4]
```diff
@@ -1,116 +1,195 @@
...
+        let (first, second) = args.two()?;
+        Ok((vec![first], second, None))
+    } else if args.len() == 3 {
+        let (first, second, third) = args.three()?;
+        Ok((vec![first], third, Some(Box::new(second))))
+    } else {
+        Err(PlanError::invalid("max_by requires 2 or 3 arguments"))
+    }?;
+    let order_by = Some(vec![expr::Sort::new(order_by, asc, false)]);
+    Ok(expr::Expr::AggregateFunction(AggregateFunction {
```
Comment (linhr) [inline crates/sail-plan/src/function/aggregate.rs:35]: Let's add a TODO here. I guess this is not efficient due to sorting.

File: crates/sail-plan/src/function/aggregate.rs [hunk 5]
```diff
@@ -1,116 +1,195 @@
...
+use crate::function::common::AggFunction;
+use crate::utils::ItemTaker;
 
 lazy_static! {
-    static ref BUILT_IN_AGGREGATE_FUNCTIONS: HashMap<&'static str, Arc<AggregateUDF>> =
+    static ref BUILT_IN_AGGREGATE_FUNCTIONS: HashMap<&'static str, AggFunction> =
         HashMap::from_iter(list_built_in_aggregate_functions());
 }
 
-struct AggregateFunctionBuilder;
```
Comment (linhr) [inline crates/sail-plan/src/function/aggregate.rs:19]: Is there any reason for removing this? I feel it's more verbose to have `_agg` in `F::default_agg`, `F::custom_agg`, etc. Also the function builder may have too much responsibility when we introduce window functions and table functions.

File: crates/sail-sql/src/expression/common.rs [hunk 6]
```diff
@@ -233,8 +156,7 @@ fn from_ast_function_arg(arg: ast::FunctionArg) -> SqlResult<spec::Expr> {
     use ast::{FunctionArg, FunctionArgExpr};
 
     match arg {
-        FunctionArg::Named { .. } => Err(SqlError::unsupported("named function argument")),
-        FunctionArg::Unnamed(arg) => {
+        FunctionArg::Unnamed(arg) | FunctionArg::Named { arg, .. } => {
```
Comment (linhr) [inline crates/sail-sql/src/expression/common.rs:160]: Let's add a TODO to support named argument names.

**Review by linhr — APPROVED** [review]
Great work! 🚀

File: crates/sail-plan/src/function/aggregate.rs (hunk 5 above)
Comment (shehabgamin) [reply crates/sail-plan/src/function/aggregate.rs:19]: Good point! I'll add it back in.

**Pull request merged**